src/toxcore/dht/server.rs

```diff
@@ -142,10 +157,14 @@ impl Server {
     }
     /// send NodesRequest to random peer at every 20 seconds
     pub fn send_nodes_req(&self, friend_pk: PublicKey) -> IoFuture<()> {
-        let mut state = self.state.write();
```

Comment:
And the same thing here as @kpp wrote - 1 write lock per method will be fine because random_node almost always is Some.

Reply:
For remove timed-out clients, create_client() has been changed.
The change is from (just create client) to (check if it exists in peers_cache).
If client already exists, it means that PingRequest() is sent and client had been created and inserted to peers_cache.
Because ping period is 60 seconds and time-out period is 122 seconds.
If every PingRequest create new client, then new client replace old client with same PK.
To do this, change happen in create_client() and to avoid deadlock send_pings() logic is changed.